code/modules/ai/ai_behaviors/human_mobs/hazard_avoidance.dm

```diff
@@ -0,0 +1,112 @@
+/datum/ai_behavior/human
+	///Mob tries to keep away from hazards
+	var/avoid_hazards = TRUE
+	///assoc list of hazards to avoid and the range to stay away from them
+	var/list/hazard_list = list()
+	///Chat lines for avoiding uncategorized hazards
+	var/list/default_avoid_chat = list("Watch out!", "Watch out, hazard!", "hazard!", "Keep away from the hazard!", "Well I've never seen a hazard like this before.")
+	///Chat lines for avoiding a live nade
+	var/list/nade_avoid_chat = list("Watch out!", "Watch out, grenade!", "Grenade!", "Run!", "Get out of the way!", "Grenade, move!")
+	///Chat lines for avoiding fire
+	var/list/fire_avoid_chat = list("Watch out!", "Watch out, fire!", "fire!", "Keep away from the fire!")
+
+/datum/ai_behavior/human/find_next_dirs()
+	. = ..()
+	if(!.)
+		return
+	if(!avoid_hazards)
+		return
+
+	var/list/dir_options = .
+	dir_options = dir_options.Copy()
+	var/list/exclude_dirs = list()
+	for(var/atom/movable/thing AS in hazard_list)
+		var/dist = get_dist(mob_parent, thing)
+		if(dist > hazard_list[thing] + 1)
+			continue
+		if(!isturf(thing.loc)) //picked up nade etc
+			continue
+		if(dist == 0)
+			if(length(dir_options)) //we want to get off the hazard, but if we're trying to go somewhere else already, then that dir is fine
+				continue
+			dir_options = CARDINAL_ALL_DIRS
+			continue
+		var/dir_to_hazard = get_dir(mob_parent, thing)
+		exclude_dirs |= dir_to_hazard
+		exclude_dirs |= turn(dir_to_hazard, 45)
+		exclude_dirs |= turn(dir_to_hazard, -45)
+
+		dir_options |= REVERSE_DIR(dir_to_hazard)
+		if(dist > (ROUND_UP(hazard_list[thing] * 0.5))) //outer half of danger zone, lets add diagonals for variation
+			dir_options |= turn(dir_to_hazard, 135)
+			dir_options |= turn(dir_to_hazard, 225)
+
+	dir_options -= exclude_dirs
+	if(length(dir_options))
+		return dir_options
+	//if hazards cause movement paralysis, we just say fuck it, and ignore them since apparently every direction is dangerous
+
+///Clear the hazard list if we change z
+/datum/ai_behavior/human/proc/on_change_z(atom/movable/source, old_z, new_z)
+	SIGNAL_HANDLER
+	for(var/hazard in hazard_list)
+		remove_hazard(hazard)
+
+///Adds a hazard to watch for
+/datum/ai_behavior/human/proc/add_hazard(datum/source, atom/hazard)
+	SIGNAL_HANDLER
+	var/turf/hazard_turf = get_turf(hazard)
+	if(hazard_turf.z != mob_parent.z)
+		return
+	var/hazard_radius = hazard.get_ai_hazard_radius(mob_parent)
+	if(isnull(hazard_radius))
+		return
+	hazard_list[hazard] = hazard_radius
+	RegisterSignals(hazard, list(COMSIG_QDELETING, COMSIG_MOVABLE_Z_CHANGED), PROC_REF(remove_hazard))
+	if(get_dist(mob_parent, hazard) > 5)
+		return
+	if(isgrenade(hazard))
+		if(prob(85))
+			try_speak(pick(nade_avoid_chat))
+		return
+	if(isfire(hazard))
+		if(prob(20))
+			try_speak(pick(fire_avoid_chat))
+		return
+
+	if(prob(20))
+		try_speak(pick(default_avoid_chat))
```

Comment:
I feel like there may be a better way to do it than hardcoding

Reply:
happy to do something if you got an idea

Reply:
I do not see the problem, the alternative would be config file ?
Hard coding allow for easy overwrite for different AI type, pretty nice